### ISSUE TITLE
AbstractGoogleClientRequest and GoogleClientRequestInitializer class not found in com.google.api-client:google-api-client-android:2.0.0

### ISSUE THREAD

**Issue by waterdrake** [googleapis/google-api-java-client#2137]
My Android app with Cloud Endpoints says these classes do not exist

```
import com.google.api.client.googleapis.services.AbstractGoogleClientRequest;
import com.google.api.client.googleapis.services.GoogleClientRequestInitializer;
```

when using any version higher than 1.31.0 of the following import

`implementation 'com.google.api-client:google-api-client-android:1.31.0'

`The latest version is [2.0.0](https://mvnrepository.com/artifact/com.google.api-client/google-api-client-android) but even 1.32.1 has the issue. Anything that is 1.31.0 or lower works.

This is what my code looks like (taken from this [link](https://github.com/GoogleCloudPlatform/android-docs-samples/blob/8fb8f5be86234a5d9cf364da11079b4496c0c0ae/endpoints-frameworks/v2/app/src/main/java/com/example/migration/endpoints/app/MainActivity.java)):

```
if (myApiService == null) {  // Only do this once
        MyApi.Builder builder = new MyApi.Builder(AndroidHttp.newCompatibleTransport(),
                new AndroidJsonFactory(), null)
                .setRootUrl("http://YOUR-PROJECT-ID.appspot.com/_ah/api/")
                .setGoogleClientRequestInitializer(new GoogleClientRequestInitializer() {
                    @Override
                    public void initialize(AbstractGoogleClientRequest<?> abstractGoogleClientRequest) throws IOException {
                        abstractGoogleClientRequest.setDisableGZipContent(true);
                    }
                });
        // end options for devappserver
        myApiService = builder.build();
    }

    try {
        return myApiService.sayHi(params[0]).execute().getData();
    } catch (IOException e) {
        return e.getMessage();
    }
```

Is there a different way to be doing this in the new versions or is something else going on?

**Comment by lqiu96** [comment]
Taking a look. Can you also send me your build.gradle? I'd like to see what versions you're importing to see if there is some version incompatibility. I see the samples repo is old and archived, so I'm going to guess you've made modifications to it to try and make it work.

**Comment by waterdrake** [comment]
Here are the pertinent imports in the working build.gradle:
```

apply plugin: 'com.android.application'
apply plugin: 'com.google.cloud.tools.endpoints-framework-client'

buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'com.google.cloud.tools:endpoints-framework-gradle-plugin:2.1.0'
    }
}

dependencies {
    implementation fileTree(dir: 'libs', include: ['*.jar'])

    implementation 'com.google.api-client:google-api-client-android:1.31.0'
    implementation 'com.google.guava:guava:31.1-jre'
    implementation 'com.google.http-client:google-http-client-android:1.42.2'
    implementation 'com.google.http-client:google-http-client-gson:1.42.2'
}
```

**Comment by lqiu96** [comment]
Ah, thanks for getting back with the `build.gradle` file! I do see the issue and it seems to have evolved from v1.31.2 -> v1.31.3.

Looks like the required dependency was removed in the version bump, which is why you can't import it in a later version after v1.31.2:
```
    <dependency>
      <groupId>com.google.api-client</groupId>
      <artifactId>google-api-client</artifactId>
    </dependency>
```
From this change: https://github.com/googleapis/google-api-java-client/pull/1703 -- which looks like to decouple the dependency imports.

CC: @suztomo made the change a while back. He might be able to give you some more color as to why it was removed.

Possible solution:
Can you try and import the missing dependency?
`implementation 'com.google.api-client:google-api-client:2.0.0'` or swap with whatever version you are looking to use

Let me know if that works. Thanks!

**Comment by suztomo** [comment]
I think I removed the googlr-api-client module from there because it wasn't needed; otherwise the compilation would fail.

Please declare the dependency that contains the class when you use the class.

**Comment by waterdrake** [comment]
@lqiu96 I tried adding the dependency you recommended like this:

```
    implementation 'com.google.api-client:google-api-client-android:2.0.0'
    implementation 'com.google.api-client:google-api-client:2.0.0'
```
    
The app compiles. However, when I actually run it and launch the android app I get this error:

> Caused by: java.lang.IllegalStateException: You are currently running with version 2.0.0 of google-api-client. You need at least version 1.15 of google-api-client to run version 1.25.0 of the MyApiService library.
>         at com.google.common.base.Preconditions.checkState(Preconditions.java:534)
>         at com.google.api.client.util.Preconditions.checkState(Preconditions.java:113)

        
This is a weird error because the version is already 2.0.0 as you can see above.

**Comment by lqiu96** [comment]
> Side note, it seems wrong to have to use both the dependencies above - since this is an android app, wouldn't just the android one be expected to work?. Regardless, just to test it out, I tried using just the `com.google.api-client:google-api-client:2.0.0` one. It compiled, but when I ran the app it gave the exact same error as above.

Personally, I agree. I would think that declaring `google-api-client-android` would have the functionality of `google-api-client`. From the linked PR above, there were some changes that have been made to help with past issues, so for now you'll have to explicitly declare both. This may be something that we could revisit in the future.

> Caused by: java.lang.IllegalStateException: You are currently running with version 2.0.0 of google-api-client. You need at least version 1.15 of google-api-client to run version 1.25.0 of the MyApiService library.
at com.google.common.base.Preconditions.checkState(Preconditions.java:534)
at com.google.api.client.util.Preconditions.checkState(Preconditions.java:113)

I've generated MyApi and see this in the generated implementation:
```
  static {
    com.google.api.client.util.Preconditions.checkState(
        com.google.api.client.googleapis.GoogleUtils.MAJOR_VERSION == 1 &&
        com.google.api.client.googleapis.GoogleUtils.MINOR_VERSION >= 15,
        "You are currently running with version %s of google-api-client. " +
        "You need at least version 1.15 of google-api-client to run version " +
        "1.25.0 of the myApi library.", com.google.api.client.googleapis.GoogleUtils.VERSION);
  }
```

It seems this is generated from the gradle plugin and I'm not entirely sure where the Major/ Minor versions are being set from. For a quick fix, can you try and use `google-api-client` v[1.35.2](https://mvnrepository.com/artifact/com.google.api-client/google-api-client/1.35.2)? We can investigate how the versions are set and see if we can include v2+ `google-apis-client`. 

CC @suztomo -- Are you familiar with the endpoints framework/ endpoints grade/ or have an idea on how these values are set?

**Comment by lqiu96** [comment]
> But I should be using the android version no? Unless there is no big difference?

If you mean to also downgrade the `-android` dependency to v1.35, you can if you want to keep both `google-api-client` and `google-api-client-android` versions in sync. iirc, there are no functionality differences between v1.35+ and v2 so having `-android` as v2+ *should* be fine.

When we get some time, we'll take a look to include v2+ versions into the preconditions so it no longer fails for the newer releases of `google-api-client`.

**Comment by waterdrake** [comment]
@lqiu96 I see. Well, I think what I will do is just keep using `'com.google.api-client:google-api-client-android:1.31.0'` alone. I don't really need to upgrade it for any features (that I know off). I will keep an eye on a new update to a new `google-api-client-android` v2+ that will work on it's own, or it may get confusing in my dependencies why I have two of a similar library.

Thank you for your help!

**Comment by lqiu96** [comment]
No problem. Let me know if you do come across any more inconsistencies or issues regarding endpoint and `google-api-client` integration. I'll take a look into generated preconditions when I get a chance.

**Comment by waterdrake** [comment]
@lqiu96 Thank you. This is another issue regarding endpoints and gradle build tools I have. But they are probably not related: https://github.com/GoogleCloudPlatform/endpoints-framework-gradle-plugin/issues/110

**Comment by waterdrake** [comment]
I am still having this issue with the 2.0 version. I think this should be re-opened.

**Comment by waterdrake** [comment]
Can this be re-opened? I can't re-open it but I can make a new issue.

**Comment by lqiu96** [comment]
Hi @waterdrake, I believe this issue is due to the endpoints framework and unfortunately we currently do not have plans to resolve it for v2+. Would you be fine with sticking with v1.x+ or migrating off the endpoints framework for app engine?

**Comment by waterdrake** [comment]
@lqiu96 Migrating off of endpoints wouldn't be a huge overhaul. Also, I'm not sure what else to migrate off too. Endpoints is it's own stand alone system as far as I know. 

As I understand, this library is required to use endpoints on Android. Unless there is a different way to access endpoints on Android I am not aware of? Here are the docs showing how to migrate to the latest cloud endpoints v2 for android: https://cloud.google.com/endpoints/docs/frameworks/java/migrating-android 

If this library doesn't work, I essentially can't use cloud endpoints on android. I could stick to v1.x+ but I worry I will be stuck with an old version in the future that may have it's own problems.

Do you have any recommendations on what other steps I could take to bring this problem visible to the Google Cloud Endpoints team? I feel their documentation is misleading if the dependencies they list are not compiling right.

Thanks again for the help.

**Comment by lqiu96** [comment]
> @lqiu96 Migrating off of endpoints wouldn't be a huge overhaul. Also, I'm not sure what else to migrate off too. Endpoints is it's own stand alone system as far as I know.
> 
> As I understand, this library is required to use endpoints on Android. Unless there is a different way to access endpoints on Android I am not aware of? Here are the docs showing how to migrate to the latest cloud endpoints v2 for android: https://cloud.google.com/endpoints/docs/frameworks/java/migrating-android
> 
> If this library doesn't work, I essentially can't use cloud endpoints on android. I could stick to v1.x+ but I worry I will be stuck with an old version in the future that may have it's own problems.
> 
> Do you have any recommendations on what other steps I could take to bring this problem visible to the Google Cloud Endpoints team? I feel their documentation is misleading if the dependencies they list are not compiling right.
> 
> Thanks again for the help.

Thanks for the context. I don't have much familiarity with Android so I currently would not know what recommend you as the best alternative. I'll do some digging and I'll ask internally for either a tool that you can use or a thread/ ticket/ issue/ repo that you could ask for more help in.

Sticking with v1.x+ isn't ideal and unfortunately the support for endpoints currently isn't in the best shape.

**Comment by lqiu96** [comment]
Hi @waterdrake, could you explain your use case/ app that uses the Endpoints framework in Android so that I can get a better picture? Endpoints Framework is a server framework and not necessary needed for app building. It may be easier if you simply remove the Endpoints Framework dependency and make the API calls with our client library.

**Comment by waterdrake** [comment]
@lqiu96 In the link https://cloud.google.com/endpoints/docs/frameworks/java/migrating-android the listed dependencies are:

compile 'com.google.http-client:google-http-client-android:1.22.0'
compile 'com.google.api-client:google-api-client:+'

From what I understand, these dependency are used to generate Android client wrappers for Endpoints. This is so you can call your API from Android using convenient auto-generated classes instead of having to manually write out HTTPS requests.

The `-android` part is what is relevant to me. In this repo, there is this folder which appears to be the android section: https://github.com/googleapis/google-api-java-client/tree/main/google-api-client-android.

As you said, Endpoints Framework is meant as a server framework. But the above `google-api-client-android` section is used to auto-generate the client side wrappers for endpoint methods for Android in Java. For some reason, updating the version to anything higher than I listed in the original post, will lead to the compiler error in Android. 

FYI, there is also an iOS repo that generates client side API classes for Endpoints: https://github.com/google/google-api-objectivec-client-for-rest

Basically, Cloud Endpoints is marketed as being able to be consumed in web, iOS and Android with auto generated native client side libraries (as explained here: https://cloud.google.com/endpoints/docs/frameworks/java/calling-from-android and also here: https://youtu.be/A7yMmvrpMJc?t=50). Not being able to do that is a huge hinderance, so I really hope this can be resolved.

**Comment by waterdrake** [comment]
Just leaving an update here. I decided to move off of these dependencies to something more new. Not sure if anyone else who stumbles upon this is having the same problem but my advice would be to move off of these dependencies as related to Cloud Endpoints.

Thanks for looking into this thus far team.
